src/processor/operator/recursive_extend/bfs_scheduler.cpp

```diff
@@ -0,0 +1,192 @@
+        return checkState;
+    } else {
+        return false;
+    }
+}
+
+/**
+ * Return value = -1: New SSSPSharedState can be started.
+ * Return value = 0, The pathLengthMorsel received did not yield any value to write to the
+ * pathLengthVector.
+ * Return value > 0, indicates pathLengths were written to pathLengthVector.
+ */
+int64_t MorselDispatcher::writeDstNodeIDAndPathLength(
+    const std::shared_ptr<FTableSharedState>& inputFTableSharedState,
+    std::vector<common::ValueVector*> vectorsToScan, std::vector<ft_col_idx_t> colIndicesToScan,
+    common::ValueVector* dstNodeIDVector, common::ValueVector* pathLengthVector,
+    common::table_id_t tableID, std::unique_ptr<BaseBFSMorsel>& baseBfsMorsel) {
+    if (!baseBfsMorsel->hasSSSPSharedState()) {
+        return -1;
+    }
+    auto ssspSharedState = baseBfsMorsel->ssspSharedState;
+    auto startScanIdxAndSize = ssspSharedState->getDstPathLengthMorsel();
+    /**
+     * [startScanIdx, Size] = [UINT64_MAX, INT64_MAX] = no work in this morsel, exit & return -1
+     * [startScanIdx, Size] = [0, -1] = SSSPSharedState was just completed (marked MORSEL_COMPLETE)
+     * Else a pathLengthMorsel has been allotted to the thread and it can proceed to write the
+     * path lengths to the pathLengthVector.
+     */
+    if (startScanIdxAndSize.first == UINT64_MAX && startScanIdxAndSize.second == INT64_MAX) {
+        return -1;
+    } else if (startScanIdxAndSize.second == -1) {
+        mutex.lock();
+        numActiveSSSP--;
+        // If all SSSP have been completed indicated by count (numActiveSSSP == 0) and no more
+        // SSSP are available indicated by state IN_PROGRESS_ALL_SRC_SCANNED then global state can
+        // be successfully changed to COMPLETE to indicate completion of SSSP Computation.
+        if (numActiveSSSP == 0 && globalState == IN_PROGRESS_ALL_SRC_SCANNED) {
+            globalState = COMPLETE;
+        }
+        mutex.unlock();
+        return -1;
+    }
+    auto size = 0u;
+    auto endIdx = startScanIdxAndSize.first + startScanIdxAndSize.second;
+    while (startScanIdxAndSize.first < endIdx) {
+        if (ssspSharedState->pathLength[startScanIdxAndSize.first] >= ssspSharedState->lowerBound) {
+            dstNodeIDVector->setValue<common::nodeID_t>(
+                size, common::nodeID_t{startScanIdxAndSize.first, tableID});
+            pathLengthVector->setValue<int64_t>(
+                size, ssspSharedState->pathLength[startScanIdxAndSize.first]);
+            size++;
+        }
+        startScanIdxAndSize.first++;
+    }
+    if (size > 0) {
+        dstNodeIDVector->state->initOriginalAndSelectedSize(size);
+        // We need to rescan the FTable to get the source for which the pathLengths were computed.
+        // This is because the thread that scanned FTable initially might not be the thread writing
+        // the pathLengths to its vector.
+        inputFTableSharedState->getTable()->scan(vectorsToScan,
```

Comment:
Instead of rescanning table. Maybe you should give the src as a parameter. Though I didn't read the parallel writing output part very carefully. I would rather take another look once you finish the write for path as well.